feat(type): Add VarcharN and VarbinaryN type support

czentgr opened the pull request:
This is a WIP PR that can be used for further discussion.
This adds basic support for being able to handle length limited data based on the type definition.

This PR also work along side the E2E test using the PrestoQueryRunner enhancement https://github.com/facebookincubator/velox/pull/10523 to test CAST expressions.

It follows mainly the VARCHARN design document: https://docs.google.com/document/d/1wsNcS3Op2E5yyW9uH28gNtzXpJqvUHE6Ds4Xi_w9w54/edit?usp=sharing

mbasmanova commented on velox/type/Type.h:
```diff
@@ -530,7 +531,7 @@ class Type : public Tree<const TypePtr>, public velox::ISerializable {
   VELOX_FLUENT_CAST(Double, DOUBLE)
-  VELOX_FLUENT_CAST(Varchar, VARCHAR)
+
   VELOX_FLUENT_CAST(Varbinary, VARBINARY)
```
Since Varchar and Varbinary types tend to go together, should we also support VarbinaryN?

czentgr replied:
The reason I had not considered VarbinaryN is because Presto does not support it. 
```
presto:default> create table varbin (col1 varbinary(10));
Query 20240823_220701_00004_uxz37 failed: line 1:22: Unknown type 'varbinary(10)' for column 'col1'
```
Of course, it doesn't mean that Velox cannot support it as a built-in type.
We can add it with the same design that we use with this PR.

mbasmanova reviewed (commented):
@czentgr Thank you for working on this.

> This is a WIP PR that can be used for further discussion.

Is there anything in particular you'd like to discuss?

> This adds basic support for being able to handle length limited data based on the type definition.

Would you expland on this and update documentation to add the new type and explain how it works: https://facebookincubator.github.io/velox/develop/types.html

In particular, it would be nice to explain why we need varchar(n) in addition to varchar. How 'n' type differs from the existing type. Why there is no varbinary(n) type?

mbasmanova commented on velox/expression/SignatureBinder.cpp:
```diff
@@ -208,6 +208,12 @@ bool SignatureBinderBase::tryBind(
   const auto& params = typeSignature.parameters();
   // Type Parameters can recurse.
   if (params.size() != actualType->parameters().size()) {
+    // Type match on binding from parameterized varchar
```
Let's make sure to clearly document this behavior.

mbasmanova commented on velox/expression/CastExpr-inl.h:
```diff
@@ -380,7 +381,17 @@ void CastExpr::applyCastKernel(
       // Write the result output to the output vector
       auto writer = exec::StringWriter<>(result, row);
-      writer.copy_from(output);
+      // For target VARCHAR cast take type length into account.
```
Let's make sure to document semantics of cast to/from the new type: https://facebookincubator.github.io/velox/functions/presto/conversion.html

mbasmanova commented on velox/type/fbhive/HiveTypeParser.cpp:
```diff
@@ -104,19 +104,20 @@ Result HiveTypeParser::parseType() {
           isPositiveInteger(length.value.toString()),
           "Varchar length must be a positive integer");
       eatToken(TokenType::RightRoundBracket);
+      return Result{VARCHAR(std::atoi(length.value.data()))};
```
Does Hive support varchar(n) type? If not, let's not add this logic.

czentgr replied:
Yes, it does support VarcharN: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Types#LanguageManualTypes-StringsstringStrings
It uses the `string` type as unlimited varchar and the `Varchar(1)` to `Varchar(65553)` as length limited varchar.

As a result the  parser understands both types. If the length token is not present it creates an length unbounded varchar.

It may not be strictly necessary to add this in this PR but I think it would lay the ground for it to work out of the box.

mbasmanova replied:
@czentgr Thank you for clarifying. It would be clearer to make this change in a separate PR, add a test, and include the above explaination in the PR description.

czentgr replied:
I'm revising my earlier comment. I believe this change is required to support VARCHARN from the beginning. When the HiveConnector is set up it compares the hive column type to the velox type used to hold the data. Both of these could come from the plan (I saw them being referenced there). And if we can't parse both, type and hive_type and generate the equivalent type interactions with the table (insert, read, etc) fail.

Example, a table with a column created as VARCHAR(20). Because the hive type is not parsed but the presto type (now parsing the length) is we get a mismatch.

```
Query 20240828_174714_00010_cnaav failed: dataType_->equivalent(*hiveType_) data type VARCHAR(20) and hive type VARCHAR do not match
```

The hive_type in this case was `VARCHAR(20)` and so we can make an exact equivalent check on type and length. But, of course, VARCHAR(20) would be compatible with (but not equivalent to unbounded VARCHAR too but we probably should try to go for a proper parse and matching.

From the presto plan for an INSERT:
```
:[{"@type":"hive-hadoop2","name":"col1","hiveType":"varchar(20)","typeSignature":"varchar(20)","hiveColumnIndex":0,"columnType":"REGULAR","requiredSubfields":[]}],"pageSinkMetadata":{"schemaTableName":{"schema":"default","table":"varchar_limit"},"table":{"databaseName":"default","tableName":"varchar_limit","owner":"czentgr","tableType":"MANAGED_TABLE","storage":{"storageFormat":{"serDe":"org.apache.hadoop.hive.ql.io.orc.OrcSerde","inputFormat":"org.apache.hadoop.hive.ql.io.orc.OrcInputFormat","outputFormat":"org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat"}
```

`"name":"col1","hiveType":"varchar(20)","typeSignature":"varchar(20)"`

mbasmanova commented on velox/vector/FlatVector.cpp:
```diff
@@ -109,6 +109,11 @@ void FlatVector<StringView>::set(vector_size_t idx, StringView value) {
     BaseVector::setNull(idx, false);
   }
+  if (type()->kind() == TypeKind::VARCHAR) {
+    VELOX_CHECK(
```
VELOX_CHECK_LE

> Value exceeds allowed Vector type size

This message is a bit vague. Would it be possible to clarify it a bit? Specifically, "Vector type" here is VARCHAR, no?

getVarcharLength seems expensive. Can we cache its value it in a member variable?

czentgr replied:
Regarding `getVarcharLength`. The base type interface does not have a function that can return parameterized members. The parameters are already a member in the type (via `std::vector<TypeParameter> parameters_;`). But the TypePtr needs casting to the actual type to get access to it (in this case the `length()` function). 
I re-used the same approach that was used for DECIMAL.

Perhaps `length` isn't the right term and we could use a common `size()` function that is added to the types? For most of the primitive types it could the implementation size and for varchar/varbinary (and eventually char) it would be the definition size rather than implementation size but that could be leading to misunderstandings?

czentgr commented on velox/vector/FlatVector.h:
```diff
@@ -596,6 +599,11 @@ class FlatVector final : public SimpleVector<T> {
+  // If this vector contains values of type length bounded VARCHAR or VARBINARY
+  // it contains the maximum allowed size of the data stored. Varchar and
+  // varbinary have the same kUnboundedLength.
+  size_t varyingTypeDataLengthMax_{VarcharType::kUnboundedLength};
```
The maximum limit is 2GB for the type itself. The check was added here to ensure we don't have a situation where we assign flat vectors where we could have implicit truncation from a a type perspective. We don't know the actual data stored. We have to make changes in the read so we don't exceed the maximum length from a data file (which is not implemented currently). 
Perhaps this check isn't really needed but the point was to avoid problems if we have a vector with column type that where the actual values exceed the type length. For example, having a vector that stores strings of length 50 (assume we use most of the defined length) and identifying the type as VARCHAR(20). But perhaps we don't have to worry about it much and remove or convert to a debug check?

I'm not sure why `StringView` would not be a good match varchar? Are you thinking of CHAR?

Yuhta replied:
Yeah I am thinking a buffer with each element occupying fixed bytes.  But if it's 2GB we cannot do that.  That being said I do see value for a special implementation for N <= 16 and StringView for N > 16.  Not sure how much this is used in real world though.

czentgr commented:
@gggrace14 Yes, I'm closing in on finalizing the current changes. Any feedback is appreciated. I'm going to push fuzzer support and parquet changes as well in separate commits on this PR (will rebase and push today).
I did indeed hope that `class VaryingLengthParametricType` can be used for CharN and attempted to design it for re-use.

I am still also updating the design doc still to keep up.

gggrace14 commented:
> @gggrace14 Yes, I'm closing in on finalizing the current changes. Any feedback is appreciated. I'm going to push expression fuzzer support and parquet changes as well in separate commits on this PR (will rebase and push today). I did indeed hope that `class VaryingLengthParametricType` can be used for CharN and attempted to design it for re-use.
> 
> I am still also updating the design doc still to keep up.

Thanks a lot, @czentgr , for moving it forward quickly! I will watch this change, and at the same time start to build some CharN pieces on top it.

mbasmanova commented:
@czentgr What's the status of this PR? How close it to being ready? What's left to do? Is it possible to break up the work into pieces to allow multiple folks to work on these?

czentgr committed:
feat(type): Add VarcharN and VarbinaryN type support

This change adds the VarcharN and VarbinaryN types to be handled
by Velox. N represents the maximum allowed data length that the column can hold.
From a function perspective Varchar and Varbinary type parameters
are unchanged. They represent the length independent type. A minor change is
made to allow signature matching of VarcharN and VarbinaryN to unbounded
varchar and varbinary respectively.
Additional changes are done to the type parser to parse the length and create the new types
and the cast function to allow truncation of the output value.

czentgr committed:
feat(fuzzer): Extend expression fuzzer to generate varying VARCHAR

This change extends the expression fuzzer to generate varying length
varchar and use the function signatures with this type.

A new option max_varchar_type_length is added that allows
customization of the maximum string length used for the type
definition and subsequent string generation.

czentgr commented:
@mbasmanova What do you think about the current implementation using a new function in the TypeCoercer to map logical types? I see that the current TypeCoercer isn't really used in the Presto path because it coerces physical type differences for the most part.

The main sticking points are really the SignatureBinder to allow different parameter sizes and the logical type check in various places because they are not handled in the type via equal/equivalent which has its own problems.
It is not as elegant as I would like.

mbasmanova commented:
@czentgr Christian, recently, we looked into this a bit more and I captured my thoughts in https://github.com/facebookincubator/axiom/pull/1184

We arrive to conclusion that there is not much value in supporting these types. It requires initial investment and ongoing maintenance for no clear benefit. We believe these are legacy times from row-wise engines (MySQL) that are not being used in modern warehouses.
